src/libsyntax_pos/span_encoding.rs

```diff
@@ -0,0 +1,163 @@
+const INLINE2_OFFSETS: [u32; 3] = [14, 13, 2];
+
+// Tag = 0b11, interned format.
+// ------------------------
+// | index 31:3 | tag 1:0 |
+// ------------------------
+const INTERNED_INDEX_SIZE: u32 = 30;
+const INTERNED_INDEX_OFFSET: u32 = 2;
+
+#[repr(C)]
+union Realign {
+    u8x4: [u8; 4],
+    u32: u32,
+}
+
+pub(crate) fn encode(sd: &SpanData) -> Span {
+    let (base, len, ctxt) = (sd.lo.0, sd.hi.0 - sd.lo.0, sd.ctxt.0);
+
+    // Can we fit the span data into this encoding?
+    let fits = |sizes: [u32; 3]| {
+        (base >> sizes[BASE_INDEX]) == 0 && (len >> sizes[LEN_INDEX]) == 0 &&
+        (ctxt >> sizes[CTXT_INDEX]) == 0
+    };
+    // Turn fields into a single `u32` value.
+    let compose = |offsets: [u32; 3], tag| {
+        (base << offsets[BASE_INDEX]) | (len << offsets[LEN_INDEX]) |
+        (ctxt << offsets[CTXT_INDEX]) | tag
+    };
+
+    let val = if fits(INLINE0_SIZES) {
+        compose(INLINE0_OFFSETS, TAG_INLINE0)
+    } else if fits(INLINE1_SIZES) {
+        compose(INLINE1_OFFSETS, TAG_INLINE1)
+    } else if fits(INLINE2_SIZES) {
+        compose(INLINE2_OFFSETS, TAG_INLINE2)
+    } else {
+        let index = with_span_interner(|interner| interner.intern(sd));
+        if (index >> INTERNED_INDEX_SIZE) == 0 {
+            (index << INTERNED_INDEX_OFFSET) | TAG_INTERNED
+        } else {
+            panic!("too many spans in a crate");
+        }
+    };
+    Span(unsafe { Realign { u32: val }.u8x4 })
+}
+
+pub(crate) fn decode(span: Span) -> SpanData {
+    let val = unsafe { Realign { u8x4: span.0 }.u32 };
+
+    // Extract a field at position `pos` having size `size`.
+    let extract = |pos, size| {
+        let mask = ((!0u32) as u64 >> (32 - size)) as u32; // Can't shift u32 by 32
+        (val >> pos) & mask
+    };
+
+    let (base, len, ctxt) = match val & TAG_MASK {
+        TAG_INLINE0 => (
+            extract(INLINE0_OFFSETS[BASE_INDEX], INLINE0_SIZES[BASE_INDEX]),
+            extract(INLINE0_OFFSETS[LEN_INDEX], INLINE0_SIZES[LEN_INDEX]),
+            extract(INLINE0_OFFSETS[CTXT_INDEX], INLINE0_SIZES[CTXT_INDEX]),
+        ),
+        TAG_INLINE1 => (
+            extract(INLINE1_OFFSETS[BASE_INDEX], INLINE1_SIZES[BASE_INDEX]),
+            extract(INLINE1_OFFSETS[LEN_INDEX], INLINE1_SIZES[LEN_INDEX]),
+            extract(INLINE1_OFFSETS[CTXT_INDEX], INLINE1_SIZES[CTXT_INDEX]),
+        ),
+        TAG_INLINE2 => (
+            extract(INLINE2_OFFSETS[BASE_INDEX], INLINE2_SIZES[BASE_INDEX]),
+            extract(INLINE2_OFFSETS[LEN_INDEX], INLINE2_SIZES[LEN_INDEX]),
+            extract(INLINE2_OFFSETS[CTXT_INDEX], INLINE2_SIZES[CTXT_INDEX]),
+        ),
+        TAG_INTERNED => {
+            let index = extract(INTERNED_INDEX_OFFSET, INTERNED_INDEX_SIZE);
+            return with_span_interner(|interner| *interner.get(index));
+        }
+        _ => unreachable!()
+    };
+    SpanData { lo: BytePos(base), hi: BytePos(base + len), ctxt: SyntaxContext(ctxt) }
```

Comment:
If encoding/decoding with 2-bit tag looks too error-prone/unmaintainable, then I can redo this with 1-bit tag instead, it will reduce the persentage of inlined spans from 82.68% to 80.01% (on rustc/libstd data).